update using-the-compiler.rst with details of optimizer cleanup sequence (#13650)
ethereum/solidity: docs/using-the-compiler.rst

              // Improve allocation of stack slots for variables, can free up stack slots early.
              // Activated by default if the Yul optimizer is activated.
              "stackAllocation": true,
              // Select optimization steps to be applied. It is also possible to modify both the
              // optimization sequence and the cleanup sequence. Instructions for each sequence
              // are separated with the : delimiter. For more information see
              // "The Optimizer > Selecting Optimizations".
              // This field is optional and if not provided, the default sequences for both
              // optimization and cleanup are used.

>               // If set to an empty value, only the default cleanup sequence is used and

Reviewer: This is a bit misleading, and ambiguous as well. If you set either the optimizer sequence or the clean-up sequence as an empty value, then no optimizer and/or cleanup sequence will be applied.
E.g. if you set the steps as `:dhfG`, no optimizer steps will be executed, whereas the clean-up sequence will be executed as provided, i.e. `dhfG`. Neither of these is default. If you however set an empty non delimited sequence, in the case the default optimizer and clean-up sequence will be executed.

Also, please change each instance of `cleanup` to `clean-up`.

Author: I have changed all the `cleanup` to `clean-up` and clarified the language as requested. Let me know what you think.

>               // This field is optional, and if not provided, the optimizer will use the default
>               // sequence if omitted. Omitting the whole field will make the compiler use
>               // defaults for both sequences but setting it to an empty value (":") will not -
>               // it will instead make the compiler use only the default cleanup sequence.

Reviewer 2: ```suggestion
              // This field is optional and if not provided, the default sequences for both
              // optimization and cleanup are used.
              // If set to an empty value, only the default cleanup sequence is used and
              // no optimization steps are applied.
```

              // Improve allocation of stack slots for variables, can free up stack slots early.
              // Activated by default if the Yul optimizer is activated.
              "stackAllocation": true,
              // Select optimization steps to be applied. It is also possible to modify both the
              // optimization sequence and the clean-up sequence. Instructions for each sequence
              // are separated with the : delimiter. For more information see
              // "The Optimizer > Selecting Optimizations".
              // This field is optional and if not provided, the default sequences for both

>               // optimization and clean-up are used. If only one of the options is provivded
>               // the other will not run unless it is also provided.

Reviewer 2: ```suggestion
              // optimization and clean-up are used. If value is provided for only one option
              // the other one will not run any steps.
```

              // If set to an empty value, only the default clean-up sequence is used and

>               // no optimization steps are applied.

Reviewer 2: [Here](https://github.com/ethereum/solidity/pull/13650#discussion_r1010246383), @nikola-matic said that, when set to an empty non delimited sequence, both optimizer and clean-up default are executed. Which one is it?

Author: So I had to remake this PR however in the original Cameel stated: "omitting the whole field will make the compiler use defaults for both sequences but setting it to an empty value will not - it will instead make the compiler use the default cleanup sequence. I.e. to clear both you set it to `:`" Which is what I was trying to convey in the text.

Original PR: https://github.com/ethereum/solidity/pull/13619

Reviewer 2: Ok. So we need to add that too. Considering the order in which we present the information, I suggest the following:
1. Field is optional
2. Field not provided, then default sequences for both optimization and clean-up
3. If one option value is empty, it will not execute, only the other one.
4. --> If both values are empty, i.e, ":", then neither optimization nor clean-up steps are executed. <-- Should be added here, IMO.
5. If value is an empty string, i.e, "", default clean-up sequence only is used.

Please, keep the text that is already present and add a sentence of your own detailing the case with delimiter `:` making both values empty.

Reviewer 2: That being said, I am thinking that we should also specify to what the first value is referring to, optimization or clean-up?

Author: Added some updates here based on your suggestions. Let me know what you think.

              "stackAllocation": true,
              // Select optimization steps to be applied.

>               // Optional, the optimizer will use the default sequence if omitted.

Reviewer: These are duplicated below.

              // Select optimization steps to be applied. It is also possible to modify both the
              // optimization sequence and the cleanup sequence. Instructions for each sequence
              // are separated with the : delimiter. For more information see
              // "The Optimizer > Selecting Optimizations".

>               // Optional, the optimizer will use the default sequence if omitted. Omitting the

Reviewer: Starting with `Optional, ...` reads a bit weird now that the above sentence has been expanded. Perhaps something like `This field is optional, and if not provided, the optimizer will ...`
